add promote command

plenarius opened the pull request.

plenarius committed:
Merge pull request #2 from plenarius/remove_claim_command

Remove claim command

plenarius committed:
Lower mwm after connecting to node

plenarius committed:
Merge pull request #3 from plenarius/lower_mwm

Lower mwm after connecting to node

plenarius committed:
add promote functionality

plenarius committed:
Merge pull request #4 from plenarius/add_promote_function

add promote functionality

plenarius committed:
Fix crash when issuing nodeinfo command with no connected node

plenarius committed:
Merge pull request #5 from plenarius/fix_nodeinfo_crash

Fix crash when issuing nodeinfo command with no connected node

plenarius committed:
Promote doesn't require seed.

plenarius committed:
Clear our prompt properly if a promote fails.

plenarius committed:
we don't need a seed or account info to replay a bundle

plenarius committed:
assume an integer is a number of history items and not a hash

plenarius committed:
introduce numAddresses command which will retrieve a set amount of addresses and its balance when retrieving account data rather than the system stopping at the next unused address after a spent address, this helps with cases where users have been sent iota to an old spent address and renders their wallet unusable due to key reuse

plenarius committed:
Add promote command to branch specifically for pull request

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+
+let promoteCount = 0;
+function interrupt() {
+    return promoteCount++ >= max_promotions;
```
Please split the increment onto a separate line.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+const max_promotions = 3;
+
+let elapsedInterval;
+let spamSeed = generateSeed()
```
semicolon

mark-bradshaw replied:
and const, unless you reassign.  Do you want to change the spam seed on each call?

plenarius replied:
This is actually a misnomer from a quick copy paste I should have cleaned up, apologies. This is for a made up address not seed. I'm not sure if we should even generate made up addresses or just decide one constant address when promoting? Like we can just use CLIAPPPROMOTION99999999.... if we want.

mark-bradshaw replied:
Here's my general rule of thumb.  Follow what the GUI wallet does.  I haven't taken a look at the code but it's an electron/javascript based app, so the code translates over pretty much directly.  Go take a peak there and see what they did.

plenarius replied:
Thanks for the tip on this btw, will use that as a guide now.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+    return promoteCount++ >= max_promotions;
+}
+// WARNING: Not cryptographically secure. Do not use any seeds generated by this generator to actually store any value.
+function generateSeed() {
```
Let's centralize seed generation.  It also happens in the seed command.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+        return callback();
+      }
+
+      var hash = args.hash;
```
No `var` please.  `const` is preferred, and `let` if necessary.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+      var hash = args.hash;
+      if (hash.length < 81) {
+        if (!data.accountData) {
+          vorpal.log(chalk.red('Please provide a full transaction hash'));
```
I don't think this error message fits the situation.

plenarius replied:
If a user has just connected and not entered a seed they can still promote so this was to catch those that may have tried `promote ABCDEF` when they've yet to enter a seed? Perhaps something like 'Promoting with a short hash is only possible with a history.' or something like that?

mark-bradshaw replied:
Yeah, I prefer that.  Since the history downloads automatically it's probably just a matter of waiting.  In similar situations (like `balance`) I've said to try the command again in a few minutes once the transaction data has finished loading.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+      }
+
+      iotajs.api.getLatestInclusion([hash], (error, states) => {
+        var confirmFound = false;
```
I prefer the filter strategy you used on line 54 above.  Let's go with something like that here again.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+          }
+        });
+        if (confirmFound)
+        {
```
Please bring the braces up on the same line as the `if`.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+        {
+          vorpal.log(chalk.red('That transaction is already confirmed.'));
+          return callback();
+        } else {
```
No need for an `else`.  The return on 77 will take care of that.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+          });
+          iotajs.api.isPromotable(hash).then(state => {
+            if (state) {
+              promoteCount = 0;
```
Need a `let` declaration here.  No `var`s down below please.

plenarius replied:
I'm a bit confused about this request. We are promoting three times, in the `interrupt` function passed into `iota.api.promoteTransaction` it increments and returns false until we hit our max. What would you suggest the best way to handle this?

mark-bradshaw replied:
I was asking that you add `let` on line 78.  It's an implied `var` now.  Do you still have questions about the promoteCount variable?  It's just a retry counter.  Is there something particular about it that is not making sense?

plenarius replied:
Sorry, I've since understood. But I see now I have messed this branch up a bit by merging my master into it after the fact so I'll clean things up and redo a PR.

plenarius replied:
Alright I'm still lost heh, apologies. I've declared the `let promoteCount` at line 9 so the `interrupt` function has access to it (otherwise I get a `ReferenceError: promoteCount is not defined`). I then set it to 0 each run of this command during a session otherwise it stays at 3 any time I try to promote a second time? How is it an implied var if I've globally declared the let or did you just not see that initial declaration? Should I not be using a global let?

mark-bradshaw replied:
Ah, I didn't see it declared on line 9.  Thanks for pointing that out.

mark-bradshaw commented on lib/commands/promote.js:
```diff
@@ -0,0 +1,119 @@
+              promoteCount = 0;
+              var params = { interrupt, delay: 1000 };
+              const transfer = generateTransfer();
+              iotajs.api.promoteTransaction(hash, data.depth, data.minWeightMagnitude, [transfer], params, function(err){
```
Let's get an arrow function here for the callback.  Man I wish we had promises.

mark-bradshaw reviewed (changes requested):
Basically good to go, just need some code style cleanup.

plenarius committed:
clean up some code style
more informative error message if trying to promote a short hash
follow ui method of 81 9s for spam address
rework checking if already confirmed

plenarius committed:
there's no option of a number as a possible argument, this I think needs to be removed from replay as well

plenarius committed:
Merge branch 'master' into pr/add_promote_function

plenarius committed:
Revert "Merge branch 'master' into pr/add_promote_function"

This reverts commit f95db6babe687dcd8b9e1e4c2fe83653d5a8a738, reversing
changes made to 4bf8ad88dcc7c891c6628daad5c54d5c694209d2.

mark-bradshaw commented:
@plenarius There's a conflict now between your branch and master.  Can you take a look at that and get it squared away?

plenarius committed:
Merge branch 'master' into pr/add_promote_function

plenarius commented:
Going to close this request and perform a new one because of a few messy mistakes.
